### ISSUE TITLE
Numpy take does not the same thing as "fancy" indexing (or I am wrongly interpreting the docs?)

### ISSUE THREAD

**Issue by jsaladich** [numpy/numpy#22881]
### Describe the issue:

Hi!
While using the amazing numpy module, I found myself in the need of replace "fancy" indexing for `numpy take`.

I love working with boolean arrays and fancy indexing but I found this combination does not works when using concretely `numpy take`.  Perhaps this is an expected beahviour, but in [Numpy's docs says](https://numpy.org/doc/stable/reference/generated/numpy.take.html): "When axis is not None, this function does the same thing as “fancy” indexing (indexing arrays using arrays)"

### Reproduce the code example:

```python
import numpy as np
# Dummy array
a = np.random.random(size=(3,5,8)) 
# Dummy boolean
b = np.asarray([False, True, False]) 

np.take(a,b, axis=0).shape == a[b,...].shape # returns False --> I would expect a True
```
```


### Error message:

_No response_

### Runtime information:

Numpy version: 1.23.4
Python version: 3.10.7 (main, Sep  7 2022, 15:22:19) [GCC 9.4.0]

### Context for the issue:

Perhaps am I wrong, but shouldn't "fancy" indexing and `np.take`behave the same with boolean arrays?

Thanks a lot for your work and your time!

**Comment by MDK8888** [comment]
Hey, I would like to work on this!

**Comment by mattip** [comment]
For those following along at home: if `np.take(a, b, axis=0)` used fancy indexing for `b`, it would return an array with `shape = (1, 5, 8)` (since b has only one `True` value). Instead, it returns the equivalent of `a[(0, 1, 0), ...].shape` which is the same as `a.shape`.

**Comment by mattip** [comment]
@MDK8888 we don't assign issues. Please go ahead and start working on it, just make sure to include #22881 in your PR description so that people who find this issue see that you've tried to fix it already. Thanks for offering to contribute here!

**Comment by seberg** [comment]
Just to note, the difference is that take doesn't do boolean indexing:
```
np.take(a,b, axis=0) == np.take(a,b.astype(np.intp), axis=0)
```

**Comment by jsaladich** [comment]
> For those following along at home: if `np.take(a, b, axis=0)` used fancy indexing for `b`, it would return an array with `shape = (1, 5, 8)` (since b has only one `True` value). Instead, it returns the equivalent of `a[(0, 1, 0), ...].shape` which is the same as `a.shape`.

Thanks for the clarification @mattip this is exactly what I would expect as a user, I guess it is also a matter of the interpretation you can do about fancy indexing and `take` . As @seberg noted right above, `take` doesn't do boolean indexing while fancy indexing does. 

In my humble opinion (perhaps skewed by my current needs), I believe it would be very useful to increase the capabilities of `take` with the addition of boolean indexing

**Comment by MDK8888** [comment]
Hey, I spent some time looking over the documentation, and while I was able to find a [link](https://github.com/numpy/numpy/blob/v1.24.0/numpy/core/fromnumeric.py#L93-L190) to the documentation for np.take, it seems that the implementation details are still a bit vague to me-I admit that I'm not too familiar with Numpy's codebase, but a link to the actual implementation would be super helpful!

**Comment by mattip** [comment]
From a quick perusal of the code, it seems there is no check whether [boolean advanced indexing](https://numpy.org/doc/stable/user/basics.indexing.html#boolean-array-indexing) should be used (is `indices` of type boolean) instead of integer array indexing. Thinking about this some more, I am not sure whether changing this would break existing code. Maybe what is needed is to document the current behavior, and maybe put this issue on the wish list of enhancements by adding a label, rather than changing the code.

But since you asked it goes like this:
- np.take can be found by searching the codebase for `def take`, which is https://github.com/numpy/numpy/blob/fe73a8498417d762a2102d759fa4c88613b398ef/numpy/core/fromnumeric.py#L94
- this then calls `getattr(obj, method)` which will look up `ndarray.take` here https://github.com/numpy/numpy/blob/fe73a8498417d762a2102d759fa4c88613b398ef/numpy/core/fromnumeric.py#L52
- ndarray is entirely built in C. So you have to find the method in the big table of `array_methods`, which maps the method to `array_take` (in general that will usually be true: a method `ndarray.foo` maps to  `array_foo`) which you can find here https://github.com/numpy/numpy/blob/fe73a8498417d762a2102d759fa4c88613b398ef/numpy/core/src/multiarray/methods.c#L110
- That in turn calls the NumPy C-API function `PyArray_TakeFrom`, which is defined here https://github.com/numpy/numpy/blob/fe73a8498417d762a2102d759fa4c88613b398ef/numpy/core/src/multiarray/item_selection.c#L188-L193, which has the actual implementation. Note there is no check there for the dtype of `indices`

**Comment by MDK8888** [comment]
@mattip, thank you so much for pointing me in the right direction! Thinking more carefully about this issue, it may be better to include it in the next release. If we allow boolean array inputs, then that boolean array must be linear-otherwise, there is no way to tell which indices we want to include when we call np.take() on an array. I'm not sure if that's the implementation that we should go with, and I wanted to discuss it further.

**Comment by rkern** [comment]
I would probably just fix the documentation to refer to "integer fancy indexing" and maybe a `Note` that it will coerce boolean index arrays to integers first.

FWIW, the history of this is that in the days of `Numeric`, numpy's predecessor, there was no fancy indexing, either of integers or booleans. We did have `take()`, though, mostly in its current form, which implemented the semantics of integer fancy indexing, a term that would only appear later when we added those semantics to `ndarray.__getitem__`. At that time, we updated the `take()` documentation to say that it implemented "fancy indexing" now that we were focusing on that concept. IIRC, only _then_ did we add the different semantics for boolean indices and lumped it under the term "fancy indexing", too.

**Comment by jsaladich** [comment]
Thanks a lot for your feedback @rkern, it is always interesting to understand the origins of what we use :)
For a potential debate, I guess numpy already has some sort of implementation to translate a boolean array into an integer indices array (e.g. `np.nonzero`). Would it be out of scope to do some kind of translation bool_arr --> int_arr respecting the indices of the `True` when using `take()`?
`take()` is already doing a translation (I guess something equivalent to `bool_arr.astype(np.intp)`) when entering a boolean array, but it doesn't seem the most adecuate

**Comment by rkern** [comment]
Yes, I think it would be out of scope. Boolean fancy indexing just doesn't fit with what `take()` does, particularly with respect to its `axis` argument. If you wanted to implement a function that implemented boolean fancy indexing, it should be a separate function, but I don't think there is much appetite for such a function. `__getitem__` works fine for most uses (you can use `operator.getitem()` if you want a function to call).

**Comment by MDK8888** [comment]
@rkern, thank you so much for helping me learn more about Numpy! I guess, then, it would be best to update the documentation rather than add a new feature.

**Comment by seberg** [comment]
Thanks for the discussion.  I agree updating the docs would be best for now.  I could also see deprecating boolean indexing if you/others feel that casting it is surprising and nobody minds, though (the error could say very clearly how to get the old behavior).
